sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs

```diff
@@ -0,0 +1,133 @@
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using System.Threading;
+
+namespace Azure.ResourceManager.Core
+{
+    /// <summary>
+    /// A class representing Azure resource manager client options.
+    /// </summary>
+    public class ApiVersions
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ApiVersions"/> class.
+        /// </summary>
+        internal ApiVersions(AzureResourceManagerClientOptions clientOptions)
+        {
+            BuildApiTable(clientOptions);
+        }
+
+        private Dictionary<string, PropertyWrapper> _loadedResourceToApiVersions = new Dictionary<string, PropertyWrapper>();
+        private Dictionary<string, string> _nonLoadedResourceToApiVersion = new Dictionary<string, string>();
+
+        private void BuildApiTable(AzureResourceManagerClientOptions clientOptions)
+        {
+            var methods = GetExtensionMethods();
+            foreach (var method in methods)
+            {
+                if (method.Name.EndsWith("RestApiVersions", StringComparison.Ordinal))
+                {
+                    var apiObject = method.Invoke(null, new object[] { clientOptions });
+                    var properties = apiObject.GetType().GetProperties();
+                    foreach (var prop in properties)
+                    {
+                        if (prop.GetValue(apiObject) is ApiVersionsBase propVal)
+                        {
+                            var key = propVal.ResourceType;
+                            _loadedResourceToApiVersions.Add(key.ToString(), new PropertyWrapper(prop, apiObject));
+                        }
+                    }
+                }
+            }
+        }
+
+        private static IEnumerable<MethodInfo> GetExtensionMethods()
+        {
+            // See TODO ADO #5692
+            var results =
+                        from assembly in AppDomain.CurrentDomain.GetAssemblies()
+                        where assembly.GetName().ToString().StartsWith("Azure.", StringComparison.Ordinal ) || assembly.GetName().ToString().StartsWith("Proto.", StringComparison.Ordinal)
+                        from type in assembly.GetTypes()
+                        where type.IsSealed && !type.IsGenericType && !type.IsNested && type.Name.Equals("AzureResourceManagerClientOptionsExtensions", StringComparison.Ordinal)
+                        from method in type.GetMethods(BindingFlags.Static | BindingFlags.Public)
+                        where method.IsDefined(typeof(ExtensionAttribute), false)
+                        where method.GetParameters()[0].ParameterType == typeof(AzureResourceManagerClientOptions)
+                        select method;
+            return results;
+        }
+
+        internal string LoadApiVersion(ProvidersOperations providers, ResourceIdentifier id, CancellationToken cancellationToken)
+        {
+            var results = providers.Get(id.Type.Namespace, null, cancellationToken);
```

Comment:
shouldn't this check to see if it is in the current list before falling back to the provider operations?

Can we add a test case which validates this?

I am also wondering if this and TryGet can get combined?

if _loadedResourceToApiVersions.TryGet
else if _nonLoadedResourceToApiVersion.TryGet
else providersOperation.Find
else return null

If we do combine them, then the signature can't have an out value since it will need an async variant

Reply:
Assumption is this is only called when the caller has already figured this out via the TryGet method or doesn't care what it is set to and wants to load it, it is internal for that reason so we control the call to it.

Possibly we can combine the TryGet here, but I don't think we want the LoadApi Network call to be called in a public method, or do we?

Or do we have mutiple TryGets? One that has the network call involved?

Reply:
you can gain some efficiency but not calling this method twice however.  So even though its internal we shouldn't need to call it more than once per client